File: app/backend/utils/dynamo/list_features.go
package dynamo

import (
	"backend/api"
	"context"
	"log"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
)

type ListFeaturesInput struct {
	LastFeatureId        *string
	ProjectionExpression *string
}

func (dc *DynamoConfig) ListFeatures(ctx context.Context, input *ListFeaturesInput) (*api.ListFeaturesResponse, error) {
	proj, exprAttrNames := sanitizeProjectionExpression(input.ProjectionExpression)
	exclusiveStartKey := makeExclusiveStartKey(input.LastFeatureId)

	queryIn := &dynamodb.QueryInput{
		TableName:                 dc.TableName,
		KeyConditionExpression:    aws.String("PK = :pk"),
		ExpressionAttributeValues: map[string]types.AttributeValue{":pk": &types.AttributeValueMemberS{Value: "FEATURE"}},
		ProjectionExpression:      proj,
		ExpressionAttributeNames:  exprAttrNames,
		ExclusiveStartKey:         exclusiveStartKey,
		ReturnConsumedCapacity:    types.ReturnConsumedCapacityTotal,
		Limit:                     aws.Int32(25),
	}
	result, err := dc.Client.Query(ctx, queryIn)
	if err != nil {
		return nil, err
	}
	features := []api.Feature{}
	err = attributevalue.UnmarshalListOfMaps(result.Items, &features)
	if err != nil {
		return nil, err
	}
	log.Printf("ConsumedCapacity: %f", *result.ConsumedCapacity.CapacityUnits)

	lastFeatureId := extractLastFeatureId(result.LastEvaluatedKey)
	return &api.ListFeaturesResponse{
		Features:      features,
		LastFeatureId: lastFeatureId,
	}, nil
}

func makeExclusiveStartKey(lastFeatureId *string) map[string]types.AttributeValue {
	if lastFeatureId == nil {
		return nil
	}
	return map[string]types.AttributeValue{
		"PK": &types.AttributeValueMemberS{Value: "FEATURE"},
		"SK": &types.AttributeValueMemberS{Value: "FEATURE#" + *lastFeatureId},
	}
}

func extractLastFeatureId(lastEvaluatedKey map[string]types.AttributeValue) *string {
	if lastEvaluatedKey == nil {
		return nil
	}
	if skAttr, ok := lastEvaluatedKey["SK"]; ok {
		if sk, ok := skAttr.(*types.AttributeValueMemberS); ok {
			parts := strings.SplitN(sk.Value, "#", 2)
			if len(parts) == 2 {
				return &parts[1]
			}
		}
	}
	return nil
}

File: app/backend/src/api_impl/dynamo/features/list_features.rs
//! List features from DynamoDB.

use anyhow::Result;
use aws_sdk_dynamodb::types::AttributeValue::{self, S};
use openapi::models::{Feature, ListFeaturesResponse};

use crate::api_impl::{
    dynamo::features::sanitize_projection_expression, mapping::try_map_item_to_feature,
};

/// Lists features with pagination and optional field projection.
///
/// Access pattern: `PK="FEATURE"`, `SK="FEATURE#{id}"`.
///
/// # Arguments
///
/// * `client` - DynamoDB client.
/// * `initial_projection_expression` - Optional projected attributes.
/// * `last_feature_id` - Optional pagination cursor.
/// * `table_name` - DynamoDB table name.
///
/// # Returns
///
/// Returns:
/// - `Ok(ListFeaturesResponse)` on success.
/// - `Err(ListFeaturesError)` on request or mapping failure.
pub async fn list_features(
    client: &aws_sdk_dynamodb::Client,
    initial_projection_expression: &Option<String>,
    last_feature_id: &Option<String>,
    table_name: impl Into<String>,
) -> Result<ListFeaturesResponse> {
    // Prevent injection attacks and ensure valid attribute names
    let (projection_expression, expression_attribute_names) =
        sanitize_projection_expression(initial_projection_expression);

    let mut query_builder = client
        .query()
        .table_name(table_name)
        .key_condition_expression("PK = :pk")
        .expression_attribute_values(":pk", S("FEATURE".to_string()))
        .set_projection_expression(projection_expression)
        .set_expression_attribute_names(expression_attribute_names)
        .limit(25); // Balance latency and throughput

    if let Some(last_id) = last_feature_id {
        query_builder = query_builder
            .exclusive_start_key("PK", S("FEATURE".to_string()))
            .exclusive_start_key("SK", S("FEATURE#".to_string() + last_id));
    }

    let query_output = query_builder.send().await?;

    let last_feature_id = query_output
        .last_evaluated_key()
        .and_then(|lek| {
            lek.get("SK")
                .and_then(|av| av.as_s().ok())
                .map(|s| s.to_string())
        })
        .map(|s| s.strip_prefix("FEATURE#").unwrap_or(&s).to_string());

    // Mapping can fail if DynamoDB data doesn't match expected schema
    let features = query_output
        .items()
        .iter()
        .map(|item| try_map_item_to_feature(item))
        .collect::<Result<Vec<Feature>, &AttributeValue>>()
        .map_err(|av| {
            anyhow::anyhow!("Data integrity error: unexpected attribute value {:?}", av)
        })?;

    Ok(ListFeaturesResponse {
        features,
        last_feature_id,
    })
}
